### PR TITLE
PYIC-7478 shared claims refactor

### PR DESCRIPTION
**Pull request by Joe-Edwards-GDS**
### What changed

Create a new class for shared claims and shared claim generation.

May be easier to review commit-by-commit

### Why did it change

The existing code is hard to follow and makes extensive use of JSON node manipulation (rather than vocab classes).

This change should make it easier to maintain and update.

### Issue tracking

- [PYIC-7478](https://govukverify.atlassian.net/browse/PYIC-7478)

[PYIC-7478]: https://govukverify.atlassian.net/browse/PYIC-7478?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/AuthorizationRequestHelper.java [hunk 1]
```diff
@@ -85,24 +83,7 @@ ResponseType.CODE, new ClientID(oauthCriConfig.getClientId()))
...
-                claimsSetBuilder.claim(SHARED_CLAIMS, response);
-            } else {
-                SharedClaimsResponseDto response =
-                        new SharedClaimsResponseDto(
-                                sharedClaims.getName(),
-                                sharedClaims.getBirthDate(),
-                                sharedClaims.getAddress());
-                claimsSetBuilder.claim(SHARED_CLAIMS, response);
-            }
+            claimsSetBuilder.claim(SHARED_CLAIMS, sharedClaims.toMapWithNoNulls());
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/AuthorizationRequestHelper.java:86]: By default Nimbus does its own JSON serialisation which ignores our Jackson annotations and include null claims.  Using Jackson to go via a map will exclude them - as with the evidence request below.

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getName() != null) {
+                    sharedClaims.getName().addAll(personWithIdentity.getName());
+                }
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getBirthDate() != null) {
+                    sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+                }
+                if (ADDRESS.equals(vc.getCri())
+                        && credentialSubject instanceof AddressAssertion addressAssertion
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java:48]: This is a small deviation from the existing implementation - previously it would add addresses from non-address VCs _until_ it found an address VC (at which point it would replace everything with that address).

Given that we always use the address from the address VC, this should be a clearer and easier way of handling it. (And if we need to support addresses on other VCs, it should be relatively clear how to do that).

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                LogHelper.buildLogMessage("Found shared claims")
+                        .with("names", sharedClaims.getName().size())
+                        .with("birthDates", sharedClaims.getBirthDate().size())
+                        .with("addresses", sharedClaims.getAddress().size())
+                        .with("emails", sharedClaims.getEmailAddress() == null ? 0 : 1)
+                        .with(
+                                "socialSecurityRecords",
+                                sharedClaims.getSocialSecurityRecord().size()));
+
+        stripDisallowedSharedClaims(sharedClaims, allowedSharedClaims);
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java:49]: It might be safer to add this into the conditions above, but then you have to add a load more null-checking in order to distinguish 'present but empty' from 'not present'.

Now its in a separate class, it should be more obvious that you need to update this at the same time...

File: libs/common-services/src/test/java/uk/gov/di/ipv/core/library/helpers/JwtHelperTest.java [hunk 4]
```diff
@@ -43,36 +36,30 @@
 
 @ExtendWith(MockitoExtension.class)
 class JwtHelperTest {
-    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    @Data
+    private static final class ExamplePayload {
```
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/test/java/uk/gov/di/ipv/core/library/helpers/JwtHelperTest.java:42]: (there's no reason for this test to use shared claims - it's just checking that we're using Nimbus correctly)

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                && personWithIdentity.getName() != null) {
+            sharedClaims.getName().addAll(personWithIdentity.getName());
+        }
+        if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                && personWithIdentity.getBirthDate() != null) {
+            sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+        }
+        if (ADDRESS.equals(vc.getCri())
+                && credentialSubject instanceof AddressAssertion addressAssertion
+                && addressAssertion.getAddress() != null) {
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java:67]: This is a small deviation from the existing implementation - previously it would add addresses from non-address VCs until it found an address VC (at which point it would replace everything with that address).

Given that we only ever expect an address from the address VC, this should be a clearer and easier way of handling it. (And if we need to support addresses on other VCs, it should be relatively clear how to do that).

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java [hunk 6]
```diff
@@ -397,70 +393,63 @@ private EvidenceRequest getEvidenceRequestForKbvCri(Vot targetVot) {
...
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getName() != null) {
+                    sharedClaims.getName().addAll(personWithIdentity.getName());
+                }
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getBirthDate() != null) {
+                    sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+                }
+                if (ADDRESS.equals(vc.getCri())
+                        && credentialSubject instanceof AddressAssertion addressAssertion
```
Comment (Wynndow) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java]: It's interesting that in the vocab classes, AddressAssertion extends Person. I thought it would just extend CredentialSubject directly...

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java [hunk 7]
```diff
@@ -397,70 +393,63 @@ private EvidenceRequest getEvidenceRequestForKbvCri(Vot targetVot) {
...
+                        && addressAssertion.getAddress() != null) {
+                    sharedClaims.getAddress().addAll(addressAssertion.getAddress());
+                }
+                // TODO: should update tests to avoid needing this
+                if (ADDRESS.equals(vc.getCri())
+                        && credentialSubject instanceof IdentityCheckSubject addressAssertion) {
+                    sharedClaims.setAddress(
+                            addressAssertion.getAddress().stream().collect(Collectors.toSet()));
```
Comment (Wynndow) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java]: Intellij is suggesting this can be simlified with:

`new HashSet<>(addressAssertion.getAddress())`

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java [hunk 8]
```diff
@@ -397,70 +393,63 @@ private EvidenceRequest getEvidenceRequestForKbvCri(Vot targetVot) {
...
+
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getName() != null) {
+                    sharedClaims.getName().addAll(personWithIdentity.getName());
+                }
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getBirthDate() != null) {
+                    sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+                }
+                if (ADDRESS.equals(vc.getCri())
```
Comment (Wynndow) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java]: Would we ever get an `AddressAssertion` _not_ from the address CRI?

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java [hunk 9]
```diff
@@ -397,70 +393,63 @@ private EvidenceRequest getEvidenceRequestForKbvCri(Vot targetVot) {
...
-    }
 
-    private String getEmailAddressFromIpvSession(
-            IpvSessionItem ipvSessionItem, List<String> allowedSharedAttr) {
-        if (ipvSessionItem.getEmailAddress() != null
-                && allowedSharedAttr.contains(SHARED_CLAIM_ATTR_EMAIL)) {
-            return ipvSessionItem.getEmailAddress();
-        }
-        return null;
+        LOGGER.info(
```
Comment (Wynndow) [inline lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java]: Should this log perhaps come after the stripping below?

File: lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java (hunk 5 above)
Comment (Wynndow) [reply lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java:67]: OK - looks like that answers my earlier question

**Review by Wynndow — APPROVED** [review]
Looks good! And that's a nice diff...
